firecloud_project.py

```diff
@@ -353,71 +381,95 @@ def generate_config(context):
   project_name = context.properties.get('projectName', project_id)
   labels_obj = context.properties.get('labels', {})
 
+  # Save this template's version number and all parameters inputs to the project metadata to keep track of what
+  # operations were performed on a project.
+  labels_obj.update({
+    "firecloud-project-template-version" : "1"
+  })
+
+  # Label text requirements include max length of 63 chars, only allowing (a-z, 0-9, -, _), and value must start with a
+  # letter. Link: https://cloud.google.com/deployment-manager/docs/creating-managing-labels#requirements
+  LABEL_MAX_LENGTH = 63
+  ALLOWED_CHARS_COMPLEMENT = r'[^a-z0-9-_]+'
+  ALLOWED_STARTING_CHARS_COMPLEMENT = r'^[^a-z]*'
+  for k, v in context.properties.items():
+    new_key = 'param--' + k.lower() # converts to lowercase
+    new_value = str(v).lower()
+    new_value = re.sub(ALLOWED_CHARS_COMPLEMENT, '--', new_value) # remove all illegal characters and replace with '--'
+    new_value = re.sub(ALLOWED_STARTING_CHARS_COMPLEMENT, '', new_value) # make sure first char is a lowercase letter
+
+    labels_obj.update({
+      new_key[0:LABEL_MAX_LENGTH] : new_value[0:LABEL_MAX_LENGTH]
+    })
+
```

Comment:
labels look like this (due to the illegal characters and 63 character limit (and some IDs were replaced)):
```
billingaccount: fc-broad-institute---123456789
firecloud-project-template-version: '1'
param--billingaccountfriendlyname: broad--institute-----123456789
param--billingaccountid: billingaccounts--123456789
param--enableflowlogs: 'true'
param--fcbillinggroup: terra-billing--test--firecloud--org
param--fcprojecteditors: serviceaccount--123456789-asdfasdfasdfasdfasdfasdf--
param--fcprojectowners: group--firecloud-project-owners--test--firecloud--org--
param--highsecuritynetwork: 'true'
param--labels: template_branch--ca-738-private-access--template_org--broadinst
param--parentfolder: ''
param--parentorganization: ''
param--privateipgoogleaccess: 'true'
param--projectid: private-access-fiab-aa41
param--projectownersgroup: policy-123asdf-123qwe-123asdf-123asdf-123asdf--dev--test--fireclo
param--projectviewersgroup: policy-123asdf-123qwe-123asdf-123asdf-123asdf--dev--test--fireclo
param--requesterpaysrole: organizations--1234567890--roles--requesterpays
template_branch: ca-738-private-access
template_org: broadinstitute
template_path: firecloud_project-py
template_repo: gcp-dm-templates
vpc-network-name: network
vpc-subnetwork-name: subnetwork
```